packages/roosterjs-editor-core/lib/coreApi/getSelectionRangeEx.ts

```diff
@@ -0,0 +1,84 @@
+import { contains, createRange, safeInstanceOf, VTable } from 'roosterjs-editor-dom';
+import {
+    EditorCore,
+    GetSelectionRangeEx,
+    SelectionRangeEx,
+    SelectionRangeTypes,
+} from 'roosterjs-editor-types';
+
+const TABLE_SELECTED = '_tableSelected';
+
+/**
+ * @internal
+ * Get current or cached selection range
+ * @param core The EditorCore object
+ * @returns A Range object of the selection range
+ */
+export const getSelectionRangeEx: GetSelectionRangeEx = (core: EditorCore) => {
+    let result: SelectionRangeEx = null;
+    if (core.lifecycle.shadowEditFragment) {
+        const shadowRange =
+            core.lifecycle.shadowEditSelectionPath &&
+            createRange(
+                core.contentDiv,
+                core.lifecycle.shadowEditSelectionPath.start,
+                core.lifecycle.shadowEditSelectionPath.end
+            );
+
+        if (shadowRange.collapsed) {
+            const tableSelected = getTableSelected(core.contentDiv);
+
+            if (tableSelected) {
+                return createTableSelectionEx(tableSelected);
+            }
+        }
+
+        return createNormalSelectionEx([shadowRange]);
+    } else {
+        if (core.api.hasFocus(core)) {
```

Comment:
If not has focus, need to get from cache.

In this API we can always assume "tryGetFromCache" to be true

Reply:
I think we have this behavior already using the selected range from the Dom plugin if no focus.